File: 74-search-a-2d-matrix/74-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        
        m, n = len(matrix), len(matrix[0])
        
        row = -1
        left, right = 0, m - 1
        while left <= right:
            mid = (left + right) // 2
            if target < matrix[mid][0]:
                right = mid - 1
            elif target > matrix[mid][-1]:
                left = mid + 1
            else:
                row = mid
                break
        
        if row == -1:
            return False
        
        left, right = 0, n - 1
        while left <= right:
            mid = (left + right) // 2
            if target < matrix[row][mid]:
                right = mid - 1
            elif target > matrix[row][mid]:
                left = mid + 1
            else:
                return True
        
        return False

File: 74-search-a-2d-matrix/74-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        
        int m = matrix.size();
        int n = matrix[0].size();
        
        int top = 0; 
        int bottom = m - 1;
        
        int row = 0;
        while(top <= bottom){
            int mid = (top + bottom) / 2;
            int start = matrix[mid][0], end = matrix[mid][n - 1];
            if(target >= start && target <= end){
                row = mid;
                break;
            }else if(target < start){
                bottom = mid - 1;
            }else{
                top = mid + 1;
            }
        }
        
        int left = 0; 
        int right = n - 1;
        
        while(left <= right){
            int mid = (left + right) / 2;
            int curr = matrix[row][mid];
            if(curr < target){
                left = mid + 1;
            }else if(curr > target){
                right = mid - 1;
            }else{
                return true;
            }
        }
        
        return false;
    }
};